### ISSUE TITLE
Limit possible reactions amount from Reacter

### ISSUE THREAD

**Issue by antonkomarev** [cybercog/laravel-love#73]
Idea by @jwktje first appeared in [Reddit comment](https://www.reddit.com/r/laravel/comments/caagsw/introducing_laravel_love_v7/et7h6h7).

> Allow someone to vote/react once per day for example.

Need to collect all feature requirements before implementation.

**Comment by jwktje** [comment]
Wouldn’t mind helping with this. What could I do to help?

**Comment by antonkomarev** [comment]
We need to have list of possible use cases for this feature to know what parts should be configurable\extendable.

**Comment by jwktje** [comment]
My use case specifically would be the following. 
To be able to build a public "suggestion box" application. People would be able to submit something, and everyone can give either thumbs up or thumbs down to the submitted ideas. This would need to work without people having to sign up.

So the Model for a "suggestion" would need to be either liked/disliked. And a visitor can only upvote/downvote any idea once per 24 hours. 

Rate limiting could maybe work via public IP.

**Comment by antonkomarev** [comment]
First of all your case requires Guest reactions which were requested before in #23. I looks like guest reactions is the most required feature for now.

If you want to implement voting without registration - there are many issues raising.
There is no way to be 100% sure all votes will be from unique users. They can use:
- Tor network to get another IP address;
- mobile network (and re-issue different IP addresses by moving from telco tower to tower);
- online web proxies;
- bot (zombie) networks.

There is another issue with IP restriction. Some of the buildings have small IP range or even one IP address for all the provider clients. If somebody already voted from such building - nobody will be able to do it again for 24 hours.

All this stuff make me think that rate limiting should be implemented on application side. Or even if it will be included in the core package - it must be extendable/overridable.

So, returning to your case, one user can vote only once per 24 hours and pick only 1 suggestion or he can vote for any amount of suggestions, but each suggestion could receive only one vote from one user per 24 hour?

Is there is an ability to "revoke" vote and get an ability to vote for another suggestion for the same 24 hours?

**Comment by jwktje** [comment]
Totally agree that ensuring a vote is from a unique voter is super hard. And almost impossible to get 100% right. Maybe a hash of the user-agent + public ip work work as a fingerprinting approach to circumvent your description of multiple users behind a public IP?

About my use-case; in any 24 hour period a person would be able to upvote/downvote every single entry. But they just can't upvote/downvote one entry more than once that day. 

Side note; Is a reddit-like approach already possible? Where you can either upvote or downvote or don't vote.

**Comment by antonkomarev** [comment]
On a first sight user-agent + IP should be better solution, but in fact it opens possibility for cheaters who can try to vote from a different browsers or write a bot which will iterate through all the variety of user-agents and trick your voting system.

To implement reddit-like approach out of the box we need to add mutually exclusive reactions. Right now it could be solved in controller on application-level. If user upvote content - you are checking if reacter downvoted this suggestion earlier, if so - remove downvote and create upvote reaction.

```php
$reacter = $user->viaLoveReacter();

if ($reacter->hasReactedTo($suggestion, 'downvote')) {
    $reacter->unreactTo($suggestion, 'downvote');
}

$reacter->reactTo($suggestion, 'upvote');
```

In case if `upvote` already exists and user trying to create `upvote` reaction one more time an exception `Cog\Contracts\Love\Reaction\Exceptions\ReactionAlreadyExists` will be thrown.

**Comment by antonkomarev** [comment]
If you need to increase amount of reactions of the same type made by same user - you need another one not yet implemented feature (like Medium Claps) which I've called Reaction Power: #26 

Each additional `upvote` will increment reaction power to +1.

In your case there is another one caveat, what will happen if user will upvote suggestion 2 times, and downvote it on the 3rd day?

**Comment by jwktje** [comment]
Okay! So if I understand correctly it would already work if I would only want either upvote or downvote, and it would be a permanent reaction per user. So no repeated upvotes every day. 

And if I would want 1 vote per 24 hours, even by guests, this is currently not possible. Right?

**Comment by antonkomarev** [comment]
Yes, got that right.

**Comment by antonkomarev** [comment]
@jwktje when #85 will be implemented users will be able to react on the same content multiple times with one reaction type. This should be enough for your use-case requirements.

**Comment by antonkomarev** [comment]
New Weighted Reaction System was merged to `dev-master` branch and will appear in upcoming v8 release.

I'm closing this issue. Feel free to continue conversation here if there is something I'm missing.
